iroh/src/node/protocol.rs

```diff
@@ -0,0 +1,118 @@
+use futures_util::future::join_all;
+use iroh_net::endpoint::Connecting;
+
+use crate::node::DocsEngine;
+
+/// Handler for incoming connections.
+pub trait Protocol: Send + Sync + IntoArcAny + fmt::Debug + 'static {
+    /// Handle an incoming connection.
+    ///
+    /// This runs on a freshly spawned tokio task so this can be long-running.
+    fn accept(self: Arc<Self>, conn: Connecting) -> BoxedFuture<Result<()>>;
+
+    /// Called when the node shuts down.
+    fn shutdown(self: Arc<Self>) -> BoxedFuture<()> {
+        Box::pin(async move {})
+    }
+}
+
+/// Helper trait to facilite casting from `Arc<dyn T>` to `Arc<dyn Any>`.
+///
+/// This trait has a blanket implementation so there is no need to implement this yourself.
+pub trait IntoArcAny {
+    fn into_arc_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync>;
+}
+
+impl<T: Send + Sync + 'static> IntoArcAny for T {
+    fn into_arc_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync> {
+        self
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub(super) struct ProtocolMap(BTreeMap<&'static [u8], Arc<dyn Protocol>>);
+
+impl ProtocolMap {
+    /// Returns the registered protocol handler for an ALPN as a concrete type.
+    pub fn get_typed<P: Protocol>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        let protocol: Arc<dyn Protocol> = self.0.get(alpn)?.clone();
+        let protocol_any: Arc<dyn Any + Send + Sync> = protocol.into_arc_any();
+        let protocol_ref = Arc::downcast(protocol_any).ok()?;
+        Some(protocol_ref)
+    }
+
+    /// Returns the registered protocol handler for an ALPN as a [`Arc<dyn Protocol>`].
+    pub fn get(&self, alpn: &[u8]) -> Option<Arc<dyn Protocol>> {
+        self.0.get(alpn).cloned()
+    }
+
+    /// Insert a protocol handler.
+    pub fn insert(&mut self, alpn: &'static [u8], handler: Arc<dyn Protocol>) {
+        self.0.insert(alpn, handler);
+    }
+
+    /// Returns an iterator of all registered ALPN protocol identifiers.
+    pub fn alpns(&self) -> impl Iterator<Item = &&[u8]> {
+        self.0.keys()
+    }
+
+    /// Shutdown all protocol handlers.
+    ///
+    /// Calls and awaits [`Protocol::shutdown`] for all registered handlers concurrently.
+    pub async fn shutdown(&self) {
+        let handlers = self.0.values().cloned().map(Protocol::shutdown);
+        join_all(handlers).await;
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct BlobsProtocol<S> {
+    rt: tokio_util::task::LocalPoolHandle,
+    store: S,
+}
+
+impl<S: iroh_blobs::store::Store> BlobsProtocol<S> {
+    pub fn new(store: S, rt: tokio_util::task::LocalPoolHandle) -> Self {
+        Self { rt, store }
+    }
+}
+
+impl<S: iroh_blobs::store::Store> Protocol for BlobsProtocol<S> {
+    fn accept(self: Arc<Self>, conn: Connecting) -> BoxedFuture<Result<()>> {
+        Box::pin(async move {
+            iroh_blobs::provider::handle_connection(
+                conn.await?,
+                self.store.clone(),
+                MockEventSender,
```

Comment:
What's the deal with this MockEventSender? If this thing is only ever used with a mock sender, maybe we can remove event sending entirely? Or is it used in iroh.network?

Reply:
I think we ripped the events out of `iroh::Node` but kept support for events in `iroh_blobs`. This is unrelated to this PR though, it just moved around.

Reply:
Yeah, I was just confused the first time I saw this because I thought you had introduced the MockEventSender... Anyway, unrelated to this PR.